### ISSUE TITLE
opts_* API problem (in Octave)

### ISSUE THREAD

**Issue by okomarov** [matlab2tikz/matlab2tikz#507]
The problem arises from:
```matlab
isequal('',[])
```
which returns `true` in Matlab and `false` in Octave (reported to Octave bug tracker on http://savannah.gnu.org/bugs/index.php?43918). 

The following code will only error in Octave, although it shouldn't: 
```matlab
opts = opts_new();
opts = opts_add(opts, 'colormap/jet');
opts = opts_add(opts, 'colormap/jet');
```

The `opts_add()` checks if the key/value pair already exists and if not then `opts_append()`s it. However, if the value to append is empty (as in the example above), no matter what its type is, the appended value will be `[]`. When we try to add the same empty option a second time, `opts_add()` recognizes the `key` as existing and compares the old value against the new one. This comparison is false in Octave and thus it issues an error. This is because `opts_add()` converts the new empty value to `char` (on line 6), but the old value stored in the `opts` is double.

I think, in **all** (public) write functions of the `opts_*` API,  we should either enforce the `char()` conversion of the value **OR** `error()` if it is not char. 

I personally prefer the second option (where empty value is converted to `''`) since I prefer to control conversions explicitly.

**Comment by egeerardyn** [comment]
@okomarov 
I see a third option to fix this, on top of the current `isequal(oldValue, newValue)`, change this to: `isequal(oldValue, newValue) || (isempty(oldValue) && isempty(newValue))`.

From what I see, this should only be done at a single place (we should leave some comments that track back to this bug report and maybe even Savannah, though). Or, we could even factor our that snippet to another function (e.g. `areEqualOrEmpty`) with the necessary comments such that this work-around is more obvious from within our codebase.

What do you think?

**Comment by okomarov** [comment]
@egeerardyn Yes, the third option is indeed viable but I don't fancy it that much since it adds complexity (to hack around Octave's bug) and leaves the API inconsistent. Since, the `opts` will ultimately be `sprintf`ed into a char, anything else should just not be allowed and we should ensure that `opts_append` adds `''` instead of `[]`. 

According to this logic, the minimal intervention would just be to have this code for both `opts_add/append`:
```matlab
    if ~exist('value','var')
        value = '';
    end
    value = char(value);
```

**Comment by egeerardyn** [comment]
@okomarov I follow your logic for our current code and so I think that your last suggestion is probably the way to go.

On a slightly related note: perhaps we should also think about the API we have here for `opts_add` and so on. Personally, I think it would be a better design if we were to pass also doubles (some values are doubles, not strings), option arrays (e.g. for marker options), cell arrays (e.g. for tick labels), ...
This is not something we should change short-term (e.g. distinguishing between cell arrays and option arrays would be a mess or impossible), but long-term if if is viable to make that part `classdef` OOP, we might want to think about the design again. Do you have an opinion on that part?

**Comment by okomarov** [comment]
@egeerardyn Good point about further development. In that case, we can create a function that infers the type and appropriately converts to char. This function can be called within `opts_add`, opts_append, ... 

If we `classdef` the API, then option arrays will be of class `opts` and thus easy to tell apart from `cellstring`. However, the new Octave release is probably a year apart, as the response to my bug report suggests.

**Comment by egeerardyn** [comment]
I think this has been fixed by #513, or at least the short term part has been. For the longer-term part, we need to overhaul the `ops_*` code so this is out of the scope of this issue report but good to go for a new one.
